File: components/backend/types/push_notifications.go
// Package types defines push notification related types
package types

import "time"

// PushSubscription represents a browser push notification subscription
type PushSubscription struct {
	Endpoint string               `json:"endpoint" binding:"required"`
	Keys     PushSubscriptionKeys `json:"keys" binding:"required"`
}

// PushSubscriptionKeys contains the encryption keys for push notifications
type PushSubscriptionKeys struct {
	P256dh string `json:"p256dh" binding:"required"`
	Auth   string `json:"auth" binding:"required"`
}

// NotificationPreferences defines which events trigger notifications
type NotificationPreferences struct {
	SessionStarted   bool `json:"sessionStarted"`
	SessionCompleted bool `json:"sessionCompleted"`
	SessionError     bool `json:"sessionError"`
	RunFinished      bool `json:"runFinished"`
	RunError         bool `json:"runError"`
}

// UserSubscription represents a user's push notification subscription for a project
type UserSubscription struct {
	ID           string                  `json:"id"`
	ProjectName  string                  `json:"projectName"`
	UserID       string                  `json:"userId"`
	Subscription PushSubscription        `json:"subscription"`
	Preferences  NotificationPreferences `json:"preferences"`
	CreatedAt    time.Time               `json:"createdAt"`
	UpdatedAt    time.Time               `json:"updatedAt"`
}

// CreateSubscriptionRequest is the request body for creating a push subscription
type CreateSubscriptionRequest struct {
	Subscription PushSubscription        `json:"subscription" binding:"required"`
	Preferences  NotificationPreferences `json:"preferences" binding:"required"`
}

// UpdatePreferencesRequest is the request body for updating notification preferences
type UpdatePreferencesRequest struct {
	Preferences NotificationPreferences `json:"preferences" binding:"required"`
}

// VapidPublicKeyResponse contains the VAPID public key for push notifications
type VapidPublicKeyResponse struct {
	PublicKey string `json:"publicKey"`
}

// PushNotificationPayload is the payload sent to the push service
type PushNotificationPayload struct {
	Title   string               `json:"title"`
	Body    string               `json:"body"`
	Icon    string               `json:"icon,omitempty"`
	Badge   string               `json:"badge,omitempty"`
	Tag     string               `json:"tag,omitempty"`
	Data    map[string]string    `json:"data,omitempty"`
	Actions []NotificationAction `json:"actions,omitempty"`
}

// NotificationAction represents an action button in a notification
type NotificationAction struct {
	Action string `json:"action"`
	Title  string `json:"title"`
	Icon   string `json:"icon,omitempty"`
}

File: components/frontend/src/services/api/push-notifications.ts
/**
 * Push Notifications API client
 * Handles browser push notification subscription management
 */

import { apiClient } from './client';

export type PushSubscription = {
  endpoint: string;
  keys: {
    p256dh: string;
    auth: string;
  };
};

export type NotificationPreferences = {
  sessionStarted: boolean;
  sessionCompleted: boolean;
  sessionError: boolean;
  runFinished: boolean;
  runError: boolean;
};

export type UserSubscription = {
  id: string;
  projectName: string;
  subscription: PushSubscription;
  preferences: NotificationPreferences;
  createdAt: string;
  updatedAt: string;
};

/**
 * Subscribe to push notifications for a project
 */
export async function subscribeToPushNotifications(
  projectName: string,
  subscription: PushSubscription,
  preferences: NotificationPreferences
): Promise<UserSubscription> {
  return apiClient.post<UserSubscription>(
    `/projects/${projectName}/push-subscriptions`,
    { subscription, preferences }
  );
}

/**
 * Update notification preferences for a subscription
 */
export async function updateNotificationPreferences(
  projectName: string,
  subscriptionId: string,
  preferences: NotificationPreferences
): Promise<UserSubscription> {
  return apiClient.put<UserSubscription>(
    `/projects/${projectName}/push-subscriptions/${subscriptionId}`,
    { preferences }
  );
}

/**
 * Unsubscribe from push notifications
 */
export async function unsubscribeFromPushNotifications(
  projectName: string,
  subscriptionId: string
): Promise<void> {
  return apiClient.delete(`/projects/${projectName}/push-subscriptions/${subscriptionId}`);
}

/**
 * Get current subscription for a project
 */
export async function getCurrentSubscription(
  projectName: string
): Promise<UserSubscription | null> {
  try {
    return await apiClient.get<UserSubscription>(
      `/projects/${projectName}/push-subscriptions/current`
    );
  } catch {
    return null;
  }
}

/**
 * Get VAPID public key for push notifications
 */
export async function getVapidPublicKey(): Promise<string> {
  const response = await apiClient.get<{ publicKey: string }>('/push-notifications/vapid-public-key');
  return response.publicKey;
}
